### DISCUSSION TITLE
Uncaught Error: Ziggy error: 'id' parameter is required for route

### DISCUSSION THREAD

**Discussion by jojogomes12** [tighten/ziggy#592] [General]
I have a problem using the route with an id
without ziggy I did like this: 
### Script.js :
```
$.ajax({
    type:'POST',

url:"{{ route('update_product',['id'=>$product->id])}}",
    data: $(this).serialize(),
    dataType:'json',
```
Now how would I do with the use of ziggy?
like this give me the error:

url:route('update_product')                                       -> 'id' parameter is required for route
url: route('update_product',['id'=>$product->id)     ->Uncaught SyntaxError: Malformed arrow function parameter list
url:route('update_product',{id : this.id})                     ->POST http://localhost:8000/products/edit 404 (Not Found)
url:route('update_product', product.id)                     ->Uncaught ReferenceError: product is not defined
url:route('update_product' ,id: $(this).serialize(id))    -> 'id' parameter is required for route

### web.php :
```
Route::get('/product/edit/{id}', 'App\Http\Controllers\ProdutosController@edit');
Route::post('/produtos/edit/{id}', 'App\Http\Controllers\ProductsController@update')->name('update_product');
```
### Controller.product :

```
public function edit($id){
    $product=Product::findOrFail($id);
   return view('product.edit',['product'=>$product]);


}

public function update(Request $request, $id){
    $product=Produto::findOrFail($id);
    $product->update([

    'Name'=>$request->name,
    'Cost'=>$request->cost,
    'Price'=>$request->price,
    'Quantity'=>$request->quantity,
    'Brands'=>$request->Brands,
      'Voltage'=>$request->Voltage,
      'Description'=>$request->Description,
    ]);

    $url = "http://localhost:8000/";

    $name['success']=false;
    $name['message']='Updated with sucess!!!';
    echo json_encode($name);


    return;
}
```

**Comment by bakerkretzmar** [comment]
It looks like you're mixing JavaScript and PHP syntax:

- `url: route('update_product')`: as the error message says, you're missing your ID parameter
- `url: route('update_product', ['id'=>$product->id])`: as the error message says, this syntax is not valid. you have a PHP array here but you're in JavaScript.
- `url: route('update_product', { id: this.id })`: this is close, what is `this.id` here?
- `url: route('update_product', product.id)`: as the error message says, there is no defined JavaScript `product` variable
- `url: route('update_product' , id: $(this).serialize(id))`: I'm not sure what `this` is or what jQuery's `serialize` does, but this is also invalid syntax I think, you've got a colon there on its own outside of an object

Based on your code sample I think this might be closer to what you want:

```blade
<script>
$.ajax({
  type: 'POST',
  url: route('update_product', {{ $produto->id }}),
  data: $(this).serialize(),
  dataType: 'json',
});
</script>
```

If this is a Blade file with a `$produto` model with ID 2, then `route('update_product', {{ $produto->id }})` should be rendered (and executed in JavaScript) as `route('update_product, 2)`, which should work.

**Reply by jojogomes12** [reply]
@bakerkretzmar the problem is javascript dont allow double keys like 
`{{  $produto->id }} `
and not allow even arrows like`->`
remember  tha im using this library :[https://github.com/tighten/ziggy](url)
to use javascript alone from blade man

**Reply by bakerkretzmar** [reply]
If this code is in a Blade template, `{{ $produto->id }}` will be rendered by Blade, so the final output of the file sent to the browser will be `route('update_product', 2)`.

If it's _not_ a Blade template, then you will need to fetch the product ID a different way.

I'm not sure what you mean by "double keys". If you can share a larger code sample and explain what you're trying to do that would be helpful.

**Reply by jojogomes12** [reply]
I'll try to exemplify the best way I can ok?

Okay so to start we have 3 separate files:
```
Main.blade.php // Loads all imports and page settings
Update.blade.php //Default update form
Script.js // All javascript functions

```
To use blade extensions we would naturally need to be inside a blade file
but with the <strong >" ziggy "</strong> extension (you can find it in the link above) you can use it as long as you have the '@routes' tag inside the file, making a very simple comparison of a file that inside the blade would look like this:
```
  {{route('function') }}
```
would be like this:
```
route(' function');
```
The problem is.. how would I use files that depend on objects to be used in this way?
for example above:
```
{{ route('update_product',['id'=>$product->id])}}
```
how can i use? inside a js file only!!!!!
If you want my complete code is on my github with the instruction on how to install it on your machine but you need to have laravel composer and a database.. 
The link: https://github.com/jojogomes12/Laravel-Crud

**Reply by bakerkretzmar** [reply]
I don't see an `update.blade.php` file in your project. Can you share a complete code example of what you're trying, and what isn't working?

I think you might still be mixing up PHP and JavaScript syntax.

This is valid Blade and will use Laravel's PHP `route()` function:

```blade
{{ route('update_product', ['id' => $product->id]) }}
```

This is valid JavaScript and will use Ziggy's JavaScript `route()` function:

```js
route('update_product', product.id);
```

This is also valid **Blade**, because it will compile to HTML. That HTML can include JavaScript inside it, which will end up using Ziggy's JavaScript `route()` function will the values passed in from Blade:

```blade
<script>
route('update_product', {{ $product->id }});
</script>
```

Note that the `{{ }}` braces are only around the product ID, not the whole line. Blade will output the product ID there and nothing else, and the route function itself will be called when the page loads.

**Reply by jojogomes12** [reply]
Oh sorry its called edit.blade.php
